README.md
# bitsdojo_window

A [Flutter package](https://pub.dev/packages/bitsdojo_window) that makes it easy to customize and work with your Flutter desktop app window **on Windows, macOS and Linux**. 

Watch the tutorial to get started. Click the image below to watch the video: 

[![IMAGE ALT TEXT](https://img.youtube.com/vi/bee2AHQpGK4/0.jpg)](https://www.youtube.com/watch?v=bee2AHQpGK4 "Click to open")

<img src="https://raw.githubusercontent.com/bitsdojo/bitsdojo_window/master/resources/screenshot.png">

**Features**:

    - Custom window frame - remove standard Windows/macOS/Linux titlebar and buttons
    - Hide window on startup
    - Show/hide window
    - Move window using Flutter widget
    - Minimize/Maximize/Restore/Close window
    - Set window size, minimum size and maximum size
    - Set window position
    - Set window alignment on screen (center/topLeft/topRight/bottomLeft/bottomRight)
    - Set window title

# Getting Started

Install the package using `pubspec.yaml`

# For Windows apps

Inside your application folder, go to `windows\runner\main.cpp` and add these two lines at the beginning of the file:

```cpp
#include <bitsdojo_window_windows/bitsdojo_window_plugin.h>
auto bdw = bitsdojo_window_configure(BDW_CUSTOM_FRAME | BDW_HIDE_ON_STARTUP);
```

# For macOS apps

Inside your application folder, go to `macos\runner\MainFlutterWindow.swift` and add this line after the one saying `import FlutterMacOS` :

```swift
import FlutterMacOS
import bitsdojo_window_macos // Add this line
```

Then change this line from:

```swift
class MainFlutterWindow: NSWindow {
```

to this:

```swift
class MainFlutterWindow: BitsdojoWindow {
```

After changing `NSWindow` to `BitsdojoWindow` add these lines below the line you changed:

```swift
override func bitsdojo_window_configure() -> UInt {
  return BDW_CUSTOM_FRAME | BDW_HIDE_ON_STARTUP
}
```

Your code should now look like this:

```swift
class MainFlutterWindow: BitsdojoWindow {
    
  override func bitsdojo_window_configure() -> UInt {
    return BDW_CUSTOM_FRAME | BDW_HIDE_ON_STARTUP
  }
    
  override func awakeFromNib() {
    ... //rest of your code
```
#

If you don't want to use a custom frame and prefer the standard window titlebar and buttons, you can remove the `BDW_CUSTOM_FRAME` flag from the code above.

If you don't want to hide the window on startup, you can remove the `BDW_HIDE_ON_STARTUP` flag from the code above.

# For Linux apps

Inside your application folder, go to `linux\my_application.cc` and add this line at the beginning of the file:

```cpp
#include <bitsdojo_window_linux/bitsdojo_window_plugin.h>
```
Then look for these two lines:

```cpp
gtk_window_set_default_size(window, 1280, 720);
gtk_widget_show(GTK_WIDGET(window));
```
and change them to this:

```cpp
auto bdw = bitsdojo_window_from(window);            // <--- add this line
bdw->setCustomFrame(true);                          // <-- add this line
//gtk_window_set_default_size(window, 1280, 720);   // <-- comment this line
gtk_widget_show(GTK_WIDGET(window));
```

As you can see, we commented the line calling `gtk_window_set_default_size` and added these two lines before `gtk_widget_show(GTK_WIDGET(window));`

```cpp
auto bdw = bitsdojo_window_from(window);
bdw->setCustomFrame(true);
```

# Flutter app integration

Now go to `lib\main.dart` and add this code in the `main` function right after `runApp(MyApp());` :

```dart
void main() {
  runApp(MyApp());

  // Add this code below

  doWhenWindowReady(() {
    const initialSize = Size(600, 450);
    appWindow.minSize = initialSize;
    appWindow.size = initialSize;
    appWindow.alignment = Alignment.center;
    appWindow.show();
  });
}
```
This will set an initial size and a minimum size for your application window, center it on the screen and show it on the screen.

You can find examples in the `example` folder.

Here is an example that displays this window:
<details>
<summary>Click to expand</summary>

```dart
import 'package:flutter/material.dart';
import 'package:bitsdojo_window/bitsdojo_window.dart';

void main() {
  runApp(const MyApp());
  doWhenWindowReady(() {
    final win = appWindow;
    const initialSize = Size(600, 450);
    win.minSize = initialSize;
    win.size = initialSize;
    win.alignment = Alignment.center;
    win.title = "Custom window with Flutter";
    win.show();
  });
}

const borderColor = Color(0xFF805306);

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: WindowBorder(
          color: borderColor,
          width: 1,
          child: Row(
            children: const [LeftSide(), RightSide()],
          ),
        ),
      ),
    );
  }
}

const sidebarColor = Color(0xFFF6A00C);

class LeftSide extends StatelessWidget {
  const LeftSide({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return SizedBox(
        width: 200,
        child: Container(
            color: sidebarColor,
            child: Column(
              children: [
                WindowTitleBarBox(child: MoveWindow()),
                Expanded(child: Container())
              ],
            )));
  }
}

const backgroundStartColor = Color(0xFFFFD500);
const backgroundEndColor = Color(0xFFF6A00C);

class RightSide extends StatelessWidget {
  const RightSide({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Container(
        decoration: const BoxDecoration(
          gradient: LinearGradient(
              begin: Alignment.topCenter,
              end: Alignment.bottomCenter,
              colors: [backgroundStartColor, backgroundEndColor],
              stops: [0.0, 1.0]),
        ),
        child: Column(children: [
          WindowTitleBarBox(
            child: Row(
              children: [Expanded(child: MoveWindow()), const WindowButtons()],
            ),
          )
        ]),
      ),
    );
  }
}

final buttonColors = WindowButtonColors(
    iconNormal: const Color(0xFF805306),
    mouseOver: const Color(0xFFF6A00C),
    mouseDown: const Color(0xFF805306),
    iconMouseOver: const Color(0xFF805306),
    iconMouseDown: const Color(0xFFFFD500));

final closeButtonColors = WindowButtonColors(
    mouseOver: const Color(0xFFD32F2F),
    mouseDown: const Color(0xFFB71C1C),
    iconNormal: const Color(0xFF805306),
    iconMouseOver: Colors.white);

class WindowButtons extends StatelessWidget {
  const WindowButtons({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        MinimizeWindowButton(colors: buttonColors),
        MaximizeWindowButton(colors: buttonColors),
        CloseWindowButton(colors: closeButtonColors),
      ],
    );
  }
}
```
</details>

#
# **Want to help? Become a sponsor**

I am developing this package in my spare time and any help is appreciated.

If you want to help you can [become a sponsor](https://github.com/sponsors/bitsdojo).

🙏 Thank you!

## ☕️ Current sponsors:

 - [Ike](https://github.com/ikeofkc)    - $5/month
 - [Cole](https://github.com/The-Funk)  - $5/month


bitsdojo_window/CHANGELOG.md
## 0.1.1+1
    - Added Linux usage instructions
## 0.1.1
    - Linux support now stable
## 0.1.0+1
    - Fix gtk library name on Linux
## 0.1.0
    - Added null safety support
## 0.0.9
    - Linux support added
## 0.0.8
    - Added macOS readme instructions
## 0.0.7
    - macOS support added
## 0.0.6
    - Works with latest Flutter version (master channel)
## 0.0.5
    - Works with latest Flutter version (dev channel)
## 0.0.4
    - Better integration with other plugins
## 0.0.3
    - Using dpi-aware values for title bar and buttons dimensions
    - Dynamically calculating default button padding instead of fixed one
## 0.0.2
    - Added video tutorial link
## 0.0.1

* Initial release
    - Custom window frame - remove standard Windows titlebar and buttons
    - Hide window on startup
    - Show/hide window
    - Minimize/Maximize/Restore/Close window
    - Move window using Flutter widget
    - Set window size, minimum size and maximum size
    - Set window position
    - Set window alignment on screen (center/topLeft/topRight/bottomLeft/bottomRight)
    - Set window title


bitsdojo_window/example/lib/switch.dart
/*
Don't forget to add these 2 lines at the beggining of windows\runner\main.cpp

#include <bitsdojo_window/bitsdojo_window_plugin.h>
auto bdw = bitsdojo_window_configure(BDW_CUSTOM_FRAME | BDW_HIDE_ON_STARTUP);

*/

import 'package:flutter/material.dart';
import 'package:bitsdojo_window/bitsdojo_window.dart';

class AppSkin {
  final Color sidebar;
  final Color backgroundStart;
  final Color backgroundEnd;
  final Color border;
  final Color buttonMouseOver;
  final Color buttonMouseDown;
  final Color icon;
  final Color iconMouseOver;
  final Color iconMouseDown;

  const AppSkin(
      {required this.sidebar,
      required this.backgroundStart,
      required this.backgroundEnd,
      required this.border,
      required this.buttonMouseOver,
      required this.buttonMouseDown,
      required this.icon,
      required this.iconMouseOver,
      required this.iconMouseDown});
}

AppSkin yellowSkin = const AppSkin(
  sidebar: Color(0xFFF6A00C),
  backgroundStart: Color(0xFFFFD500),
  backgroundEnd: Color(0xFFF6A00C),
  border: Color(0xFF805306),
  buttonMouseOver: Color(0xFFF6A00C),
  buttonMouseDown: Color(0xFF805306),
  icon: Color(0xFF805306),
  iconMouseOver: Color(0xFF805306),
  iconMouseDown: Color(0xFFFFD500),
);

AppSkin greenSkin = const AppSkin(
    sidebar: Color(0xFF198A00),
    backgroundStart: Color(0xFF25C901),
    backgroundEnd: Color(0xFF198A00),
    border: Color(0xFF0C4500),
    buttonMouseOver: Color(0xFF198A00),
    buttonMouseDown: Color(0xFF0C4500),
    icon: Color(0xFF0C4500),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

AppSkin blueSkin = const AppSkin(
    sidebar: Color(0xFF125CDB),
    backgroundStart: Color(0xFF079BF2),
    backgroundEnd: Color(0xFF125CDB),
    border: Color(0xFF092E6E),
    buttonMouseOver: Color(0xFF125CDB),
    buttonMouseDown: Color(0xFF092E6E),
    icon: Color(0xFFFFFFFF),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

AppSkin purpleSkin = const AppSkin(
    sidebar: Color(0xFF8700B2),
    backgroundStart: Color(0xFFCC00C5),
    backgroundEnd: Color(0xFF9A00CC),
    border: Color(0xFF730099),
    buttonMouseOver: Color(0xFF9A00CC),
    buttonMouseDown: Color(0xFF730099),
    icon: Color(0xFFFFFFFF),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

AppSkin cherrySkin = const AppSkin(
    sidebar: Color(0xFF850250),
    backgroundStart: Color(0xFFC90078),
    backgroundEnd: Color(0xFF850250),
    border: Color(0xFF700043),
    buttonMouseOver: Color(0xFF850250),
    buttonMouseDown: Color(0xFF700043),
    icon: Color(0xFFFFFFFF),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

final skins = [yellowSkin, greenSkin, blueSkin, purpleSkin, cherrySkin];

class AppColors extends InheritedWidget {
  final AppSkin colors;
  //final Widget child;
  const AppColors({
    Key? key,
    required this.colors,
    required Widget child,
  }) : super(key: key, child: child);

  static AppColors? of(BuildContext context) {
    return context.dependOnInheritedWidgetOfExactType<AppColors>();
  }

  @override
  bool updateShouldNotify(AppColors oldWidget) => colors != oldWidget.colors;
}

void main() {
  runApp(const MyApp());
  doWhenWindowReady(() {
    const initialSize = Size(600, 450);
    final win = appWindow;
    win.alignment = Alignment.center;
    win.minSize = initialSize;
    win.size = initialSize;
    win.title = "Switch colors Flutter app";
    win.show();
  });
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
        debugShowCheckedModeBanner: false, home: Scaffold(body: AppBody()));
  }
}

class AppBody extends StatefulWidget {
  const AppBody({Key? key}) : super(key: key);

  @override
  _AppBodyState createState() => _AppBodyState();
}

class _AppBodyState extends State<AppBody> {
  int _skinNumber = 0;
  AppSkin _skin = skins[0];

  void switchSkin() {
    _skinNumber++;
    _skinNumber = (_skinNumber % skins.length);
    setState(() {
      _skin = skins[_skinNumber];
    });
  }

  @override
  Widget build(BuildContext context) {
    return AppColors(
        colors: _skin,
        child: WindowBorder(
            color: _skin.border,
            width: 1,
            child: Row(children: [
              const LeftSide(),
              RightSide(
                onButtonPressed: switchSkin,
              )
            ])));
  }
}

class LeftSide extends StatelessWidget {
  const LeftSide({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final colors = AppColors.of(context)!.colors;
    return SizedBox(
        width: 200,
        child: Container(
            color: colors.sidebar,
            child: Column(
              children: [WindowTitleBarBox(child: MoveWindow())],
            )));
  }
}

class RightSide extends StatelessWidget {
  final VoidCallback? onButtonPressed;
  const RightSide({Key? key, this.onButtonPressed}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    final colors = AppColors.of(context)!.colors;
    return Expanded(
        child: Container(
            decoration: BoxDecoration(
              gradient: LinearGradient(
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                  colors: [colors.backgroundStart, colors.backgroundEnd],
                  stops: const [0.0, 1.0]),
            ),
            child: Column(children: [
              const RightSideTopArea(),
              Expanded(
                child: Center(
                    child: RoundedFlatButton(
                        text: 'Switch',
                        onPressed: () {
                          if (onButtonPressed != null) onButtonPressed!();
                        })),
              )
            ])));
  }
}

class RightSideTopArea extends StatelessWidget {
  const RightSideTopArea({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return WindowTitleBarBox(
        child: Row(
            children: [Expanded(child: MoveWindow()), const WindowButtons()]));
  }
}

class WindowButtons extends StatelessWidget {
  const WindowButtons({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final colors = AppColors.of(context)!.colors;
    final buttonColors = WindowButtonColors(
        iconNormal: colors.icon,
        iconMouseDown: colors.iconMouseDown,
        iconMouseOver: colors.iconMouseOver,
        mouseDown: colors.buttonMouseDown,
        mouseOver: colors.buttonMouseOver);
    final closeButtonColors = WindowButtonColors(
        mouseOver: Colors.red[700],
        mouseDown: Colors.red[900],
        iconNormal: colors.icon,
        iconMouseOver: Colors.white);

    return Row(
      children: [
        MinimizeWindowButton(colors: buttonColors),
        MaximizeWindowButton(colors: buttonColors),
        CloseWindowButton(colors: closeButtonColors),
      ],
    );
  }
}

class RoundedFlatButton extends StatelessWidget {
  final void Function()? onPressed;
  final Color? color;
  final Color? mouseOverColor;
  final Color? textColor;
  final String? text;

  const RoundedFlatButton(
      {Key? key,
      this.onPressed,
      this.color,
      this.mouseOverColor,
      this.textColor,
      this.text})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      child: Text(text ?? 'Button'),
      style: ElevatedButton.styleFrom(
          shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(12.0),
              side: const BorderSide(color: Colors.white)),
          primary: color ?? Colors.grey[900],
          padding: const EdgeInsets.symmetric(horizontal: 40, vertical: 20),
          textStyle: TextStyle(
              color: textColor ?? Colors.white,
              fontSize: 12.0,
              fontWeight: FontWeight.w600)),
      onPressed: onPressed,
    );
  }
}
